### PR TITLE
Ssl support implementation

### PR DESCRIPTION
**Pull request by ligu**
Please review whether the implementation is aligned with Your ideas.
Successfully tested on my local workstation.

Example codes for implementing the authentication related delegates in ISslConfiguration:
https://msdn.microsoft.com/en-us/library/system.net.security.localcertificateselectioncallback.aspx
https://msdn.microsoft.com/en-us/library/system.net.security.remotecertificatevalidationcallback.aspx

### CODE REVIEW

**Comment by awr** [comment]
Not sure how I missed this before -- I apologize.
I'll take a look shortly.

File: src/KafkaClient/ProducerExtensions.cs [hunk 1]
```diff
@@ -31,7 +31,7 @@ public static Task<ImmutableList<ProduceResponse.Topic>> SendMessagesAsync(this
         /// <param name="partition">The partition to send messages to</param>
         /// <param name="cancellationToken"></param>
         /// <returns>List of ProduceTopic response from each partition sent to or empty list if acks = 0.</returns>
-        public static Task<ImmutableList<ProduceResponse.Topic>> SendMessagesAsync(this IProducer producer, IEnumerable<Message> messages, string topicName, int partition, CancellationToken cancellationToken)
+        public static Task<ImmutableList<ProduceResponse.Topic>> SendMessagesAsync(this IProducer producer, IEnumerable<Message> messages, string topicName, int? partition, CancellationToken cancellationToken)
```
Comment (awr) [inline src/KafkaClient/ProducerExtensions.cs]: Is this change necessary?

File: src/KafkaClient/Connections/TcpSocket.cs [hunk 2]
```diff
@@ -247,14 +251,38 @@ private Exception WrappedException(Exception ex)
...
-        private async Task<NetworkStream> GetStreamAsync()
+        private async Task<Stream> GetStreamAsync()
         {
             using (await _clientLock.LockAsync(_disposeToken.Token).ConfigureAwait(false)) {
                 if ((_client == null || _client.Connected == false) && !_disposeToken.IsCancellationRequested) {
                     _client = await ReEstablishConnectionAsync().ConfigureAwait(false);
                 }
 
-                return _client?.GetStream();
+                if(_sslConfiguration == null || _client == null)
```
Comment (awr) [inline src/KafkaClient/Connections/TcpSocket.cs]: We'll need some tests around this new functionality.

File: src/KafkaClient/Connections/ISslConfiguration.cs [hunk 3]
```diff
@@ -0,0 +1,13 @@
+﻿using System.Net.Security;
+
+namespace KafkaClient.Connections
+{
+    public interface ISslConfiguration
```
Comment (awr) [inline src/KafkaClient/Connections/ISslConfiguration.cs:5]: Is there a simple, default implementation that easily enables ssl? That'd be super helpful for getting started.

File: src/KafkaClient/KafkaOptions.cs [hunk 4]
```diff
@@ -49,6 +52,11 @@ public class KafkaOptions
         public IConnectionConfiguration ConnectionConfiguration { get; }
 
         /// <summary>
+        /// Configuration for SSL encrypted communication
+        /// </summary>
+        public ISslConfiguration SslConfiguration { get; }
```
Comment (awr) [inline src/KafkaClient/KafkaOptions.cs]: Since ssl is part of the connection, what about this being a property of the IConnectionConfiguration? That should also help to reduce the number of places we need to pass this in.

Can you think of a reason we'd want to keep them separate, or change them independently?

File: src/KafkaClient/ProducerExtensions.cs (hunk 1 above)
Comment (awr) [inline src/KafkaClient/ProducerExtensions.cs]: Not sure I understand this change.

File: src/KafkaClient/Connections/TcpSocket.cs (hunk 2 above)
Comment (awr) [inline src/KafkaClient/Connections/TcpSocket.cs]: We'll need some tests for this functionality as well.

**Review by awr — CHANGES_REQUESTED** [review]
Overall, this is great -- thanks! A couple things I would like to get updated before taking this onboard. Let me know what your thoughts are.

File: src/KafkaClient/ProducerExtensions.cs (hunk 1 above)
Comment (ligu) [reply src/KafkaClient/ProducerExtensions.cs]: Me neither.
Most probably this was some accidental typing, You can remove it.

File: src/KafkaClient/Connections/TcpSocket.cs (hunk 2 above)
Comment (ligu) [reply src/KafkaClient/Connections/TcpSocket.cs]: Agreed.

File: src/KafkaClient/KafkaOptions.cs (hunk 4 above)
Comment (ligu) [reply src/KafkaClient/KafkaOptions.cs]: This is a good idea, I'll do the suggested change.

File: src/KafkaClient/Connections/ISslConfiguration.cs (hunk 3 above)
Comment (ligu) [reply src/KafkaClient/Connections/ISslConfiguration.cs:5]: There is no really good way to have a default implementation, because the client-side ssl setup must match with the server-side settings. For security reasons it's advised to check certificate properties, (like the owner/ussuer is matches what You expect etc.), and obviously there are no default values for these.
I've still created BasicSslConfiguration to have en example implementation.

**Comment by ligu** [comment]
Pls review my updates!
I can do some unit tests also, If You agree with these changes.

**Comment by awr** [comment]
Looking great -- thanks for updating! If you don't mind adding in some unit tests, I'd gladly merge it in.

**Review by awr — APPROVED** [review]

**Comment by ligu** [comment]
Dont merge yet, my tests shows reconnection issues! Will push the fixes later.

**Comment by awr** [comment]
Thanks for the update -- holding off for now :)

**Comment by ligu** [comment]
I also wanted to merge the SSL implementation against Your latest changes from master branch, but could not.
You've removed TcpClient, and used Socket directly for network communication.
Unfortunately SSL support could not be built on top of Socket, because it needs NetworkStream.

If You don't wish to use TcpClient, we can keep using the Socket approach, but network reading/writing should happen through a Stream, not through Socket directly. That can be NetworkStream for unencrypted, and replaced by SslStream, for encrypted connections. The NetworkStream can be easily created from the Socket, and we should keep the Stream reference in the Connection class, with/instead of Socket.

**Comment by awr** [comment]
I absolutely missed the fact that the ssl implementation required TcpClient when I was doing the performance work -- I'm trying to tie something off wrt the consumer groups right now, then I'll take a look at reintroducing Streams into the Connection (so we can use the SslStream like you suggested).

**Comment by ligu** [comment]
I'm trying to do the reintroducing of Streams. SSL support is pretty urgent for us, and I could not found any other lib supporting it. I'll come up with a PR soon.

**Comment by awr** [comment]
Not sure how far you got with this, but I've finally made progress on the groups side (which was my highest priority). I've got a bit of cleanup to do as a result, then I can come back and look at this -- likely get to it Monday if you don't beat me to it.

You mentioned earlier that you had tests that showed issues (pre merge with master). If you've got those tests (and fixes) handy, can you push them to the PR? If you do, I can help with the merging (and reintroduction of streams) to leverage your work.

**Comment by ligu** [comment]
Here is a solution for re-introducing streams, and adding SSL support.
https://github.com/ligu/KafkaClient/tree/streams
(Latest changes from You master repo added, altough the tests are not fixed yet.)
I've separated the low-level transport-related functionality from Connection, and created an ITransport interface, what Connection can use to do the low-level network communication. There is the StreamTransport implementation for the interface. We can also have a SocketTransport implementation (without the SSL support) if You think it can give advantages.

**Comment by ligu** [comment]
About the issues.
What I've noticed is that in case of the TCP connection goes down, (which is quite common after a couple of seconds, if You don't have keep-alive configured), then upon trying to publish the next message, it fails, and that message get lost. I'm still investigating this issue, no solution yet.
One idea is to add the TCP keep-alive functionality on a configurable way. But that's just a partial solution, still doesn't solve the message loosing problem in case of lost connection.

**Comment by awr** [comment]
The code in https://github.com/ligu/KafkaClient/tree/streams generally looks good to me -- I like the split with ITransport. There may be something we need to do for reconnection attempts inside the semaphore, but a bit hard to read fully on github. As for the message lost issue, if we can identify that there's a partial read, the comment in Connection: `// when reconnecting, will the data continue to stream or start again??` would be the place to distinguish between lost or partially read. If the message is actually lost, the request will timeout and cleanup appropriately otherwise.

**Comment by awr** [comment]
Since I've made changes that (again) break the work you've done, I'm going to go through the merge. I don't want to lose the work that you've done. I'll pull from your updated branch and resolve conflicts on my side.

**Comment by awr** [comment]
Closing this in favour of #22, to get the code merged in.
